landingai/vision_pipeline.py

```diff
@@ -0,0 +1,338 @@
+        self._cap = cap
+        self._lock = threading.Lock()
+        self._t = threading.Thread(target=self._reader)
+        self._t.daemon = True
+        self._t.start()
+
+    def __del__(self) -> None:
+        self._cap.release()
+
+    # grab frames as soon as they are available
+    def _reader(self) -> None:
+        while True:
+            with self._lock:
+                ret = self._cap.grab()
+                time.sleep(self._FPS)  # Limit acquisition speed
+            if not ret:
+                raise Exception(f"Connection to camera broken ({self.stream_url})")
+
+    # retrieve latest frame
+    def get_latest_frame(self) -> "FrameSet":
+        """Return the most up to date frame by dropping all by the latest frame. This function is blocking"""
+        with self._lock:
+            if self.capture_interval is not None:
+                t = datetime.now()
+                delta = (t - self._last_capture_time).total_seconds()
+                if delta <= self.capture_interval:
+                    time.sleep(delta)
+                self._last_capture_time = t
+
+            ret, frame = self._cap.retrieve()
+            if not ret:
+                raise Exception(f"Connection to camera broken ({self.stream_url})")
+            if self.motion_detection_threshold > 0:
+                if self._detect_motion(frame):
+                    return FrameSet.from_array(frame)
+                else:
+                    return FrameSet()  # Empty frame
+
+        return FrameSet.from_array(frame)
+
+    def _detect_motion(self, frame: np.ndarray) -> bool:
+        """ """
+        # Prepare image; grayscale and blur
+        prepared_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+        prepared_frame = cv2.GaussianBlur(src=prepared_frame, ksize=(5, 5), sigmaX=0)
+
+        if self.previous_frame is None:
+            # Save the result for the next invocation
+            self.previous_frame = prepared_frame
+            return True  # First frame; there is no previous one yet
+
+        # calculate difference and update previous frame TODO: don't assume the processed image is cached
+        diff_frame = cv2.absdiff(src1=self.previous_frame, src2=prepared_frame)
+        # Only take different areas that are different enough (>20 / 255)
+        thresh_frame = cv2.threshold(
+            src=diff_frame, thresh=20, maxval=255, type=cv2.THRESH_BINARY
+        )[1]
+        change_percentage = (
+            100 * cv2.countNonZero(thresh_frame) / (frame.shape[0] * frame.shape[1])
+        )
+        # print(f"Image change {change_percentage:.2f}%")
+        if change_percentage > self.motion_detection_threshold:
+            self.previous_frame = prepared_frame
+            return True
+        return False
+
+    # Make the class iterable
+    def __iter__(self) -> Any:
+        return self
+
+    def __next__(self) -> "FrameSet":
+        return self.get_latest_frame()
```

Comment:
i think frame iteration is also worth a test

Reply:
Agree. Added a TODO